### ISSUE TITLE
Unions with Zod (like allowing empty string) don't show custom error messages, always Invalid input

### ISSUE THREAD

**Issue by verekia** [react-hook-form/resolvers#167]
**Describe the bug**

- React Hook Form sends empty fields as `''` empty strings.
- Zod can only validate empty strings using unions (see [this issue](https://github.com/colinhacks/zod/issues/310)), such as using the `or` union shorthand: `z.string().url().or(z.literal(""))`, which works correctly.
- The problem arises when using custom error messages: `z.string().url('bad URL').or(z.literal(""))` always shows the generic `Invalid input` message.

This basically makes it impossible to use React Hook Forms + Zod + optional fields + custom messages.

**To Reproduce**

1. Go to the [Sandbox](https://codesandbox.io/s/lively-smoke-dm61w?file=/src/App.js)
2. Write `abc` in the field and submit
3. See `Invalid input` error instead of the expected `bad URL`

**Codesandbox link**
https://codesandbox.io/s/lively-smoke-dm61w?file=/src/App.js

**Expected behavior**
The Zod resolver should show the correct custom message.

**Desktop (please complete the following information):**

- OS: MacOS 11.2.3
- Browser Brave
- Version 1.23.73

**Investigation**

The structure of errors is deeper when a union is used. See the difference:
`z.string().url('bad URL').parse('abc')`
```
ZodError: [
  {
    "validation": "url",
    "code": "invalid_string",
    "message": "bad URL",
    "path": []
  }
]
```
vs union:
`z.string().url('bad URL').or(z.literal('')).parse('abc')`
```
ZodError: [
  {
    "code": "invalid_union",
    "unionErrors": [
      {
        "issues": [
          {
            "validation": "url",
            "code": "invalid_string",
            "message": "bad URL",
            "path": []
          }
        ]
      },
      {
        "issues": [
          {
            "code": "invalid_type",
            "expected": "",
            "received": "abc",
            "path": [],
            "message": "Expected , received abc"
          }
        ]
      }
    ],
    "path": [],
    "message": "Invalid input"
  }
]
```

**Comment by jorisre** [comment]
Good catch @verekia!

Currently, when `criteriaMode: 'firstError'` _(default),_ we 'map' over zod errors and format the first one, that's why you get:
```ts
{
  "type": "invalid_union",
  "message": "Invalid input"
}
```

That means, we have to take the first error of the union errors, something like that:
```ts
{
  message: error.unionErrors[0].errors[0].message,
  type: error.unionErrors[0].errors[0].code,
}
```
What are your thoughts?

***

You can also set `criteriaMode: 'all'` and get all errors:
![Screenshot 2021-05-13 at 21 03 20](https://user-images.githubusercontent.com/7545547/118174072-a8e4e300-b42e-11eb-8386-2b900e38aac4.png)

**Comment by verekia** [comment]
Taking the first union error seems good to me :)

In the meantime I made a user-land workaround using `criteriaMode: 'all'`. I didn't know about that option, thank you :)

https://codesandbox.io/s/optimistic-bhaskara-tx2ei?file=/src/App.js

```js
const getFixedErrors = (errors) => {
  const fixedErrors = {}
  Object.entries(errors).forEach(([errorKey, error]) => {
    const fixedError = error
    if (error.type === "invalid_union") {
      const [type, message] = Object.entries(error.types).find(
        ([messageKey]) => messageKey !== "invalid_union"
      )
      fixedError.message = Array.isArray(message) ? message[0] : message
      fixedError.type = type
    }
    fixedErrors[errorKey] = fixedError
  })
  return fixedErrors
}
```

**Comment by jorisre** [comment]
Seems good, I'll open a PR for this improvement

@bluebill1049 is that ok for you?

**Comment by bluebill1049** [comment]
Sounds good to me @jorisre thanks!

**Comment by jorisre** [comment]
@verekia fixed in the 2.5.1, let me know if you've an issue

**Comment by verekia** [comment]
Looks solved to me! Thank you very much for the super fast fix.
